src/util/memory.cpp

```diff
@@ -177,4 +177,150 @@ std::vector<std::pair<uint32_t, uint32_t>> getDirtyRegions(const uint8_t* ptr,
 
     return regions;
 }
+
+// -------------------------
+// Allocation
+// -------------------------
+
+uint8_t* allocateSharedMemory(size_t size)
+{
+    if (size < HOST_PAGE_SIZE) {
+        SPDLOG_WARN("Allocating less than a page of memory");
+    }
+
+    void* mmapRes =
+      ::mmap(nullptr, size, PROT_WRITE, MAP_SHARED | MAP_ANONYMOUS, -1, 0);
```

Comment:
The protection flags should most likely be `PROT_READ|PROT_WRITE` here. While x86 doesn't have write-only mappings, so r+w is implicitly used, this is not what this function is requesting. Same for other mmap calls below